File: hackathon/security-copilot/backend/app/schemas/article.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict


class ArticleResponse(BaseModel):
    id: int
    source_id: Optional[int] = None
    title: str
    url: str
    summary: Optional[str] = None
    author: Optional[str] = None
    published_at: Optional[datetime] = None
    collected_at: datetime
    status: str
    severity: Optional[str] = None
    trust_score: int
    image_url: Optional[str] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class ArticleListResponse(BaseModel):
    items: list[ArticleResponse]
    total: int
    skip: int
    limit: int


class VerificationResultResponse(BaseModel):
    id: int
    article_id: int
    authenticity_score: int
    credibility_score: int
    severity: Optional[str] = None
    confidence: Optional[str] = None
    business_impact: Optional[str] = None
    ai_analysis: Optional[str] = None
    trust_score_breakdown: Optional[Any] = None
    trust_level: Optional[str] = None
    cve_references: Optional[Any] = None
    sources_checked: Optional[Any] = None
    affected_products: Optional[Any] = None
    recommended_actions: Optional[str] = None
    verified_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ArticleDetailResponse(ArticleResponse):
    content: Optional[str] = None
    verification_result: Optional[VerificationResultResponse] = None
    approvals: list[Any] = []
    generated_content: list[Any] = []


# Resolve forward references with concrete types at module level
# Must run after approval.py and content.py are importable
def _rebuild():
    from app.schemas.approval import ApprovalResponse
    from app.schemas.content import GeneratedContentResponse

    ArticleDetailResponse.__annotations__["approvals"] = list[ApprovalResponse]
    ArticleDetailResponse.__annotations__["generated_content"] = list[GeneratedContentResponse]
    ArticleDetailResponse.model_rebuild(force=True)


_rebuild()

File: hackathon/security-copilot/frontend/src/types/article.ts
import type { Approval } from './approval';

export enum ArticleStatus {
  New = 'new',
  AiVerified = 'ai_verified',
  PendingManualReview = 'pending_manual_review',
  Approved = 'approved',
  Published = 'published',
  Rejected = 'rejected',
  UnderReview = 'under_review',
}

export enum ArticleSeverity {
  Critical = 'critical',
  High = 'high',
  Medium = 'medium',
  Low = 'low',
  Info = 'info',
}

export interface Article {
  id: number;
  source_id: number | null;
  title: string;
  url: string;
  content: string | null;
  summary: string | null;
  author: string | null;
  published_at: string | null;
  collected_at: string;
  status: ArticleStatus;
  severity: ArticleSeverity | null;
  trust_score: number;
  content_hash: string | null;
}

export interface VerificationResult {
  id: number;
  article_id: number;
  authenticity_score: number;
  credibility_score: number;
  severity: ArticleSeverity | null;
  confidence: 'high' | 'medium' | 'low' | null;
  business_impact: string | null;
  ai_analysis: string | null;
  trust_score_breakdown: Record<string, number> | null;
  cve_references: string[] | null;
  sources_checked: Array<{ name: string; found: boolean }> | null;
  affected_products: string[] | null;
  recommended_actions: string | null;
  verified_at: string;
}

// Re-export Approval so consumers can import from this module
export type { Approval };

export interface GeneratedContentSummary {
  id: number;
  article_id: number;
  content_type: string;
  title: string | null;
  content: string;
  is_approved: boolean;
  created_at: string;
  updated_at: string;
}

export interface ArticleDetail extends Article {
  verification_result: VerificationResult | null;
  approvals: Approval[];
  generated_content: GeneratedContentSummary[];
}

export interface ArticleListResponse {
  items: Article[];
  total: number;
  skip: number;
  limit: number;
}

export interface ArticleListParams {
  status?: ArticleStatus;
  severity?: ArticleSeverity;
  source_id?: number;
  search?: string;
  skip?: number;
  limit?: number;
}
